page/multi page properties

Menefrego opened the pull request.

Commit:
page/multi page properties

Commit:
repeater draft

dsyrstad commented on vsc_pdf_template_transformer/lib/models/tpl_repeater.dart:
```diff
@@ -0,0 +1,33 @@
+  String className = 'TplRepeater';
+
+  @WidgetJsonConverter()
+  List<wb.WidgetBuilder?>? array;
```
`array` is data only. It provides the array of data to be iterated over. The `childTemplate` will be built for each element  in the array. Hence, this should be `dynamic` and be evaluated - see below.

dsyrstad commented on vsc_pdf_template_transformer/lib/models/tpl_repeater.dart:
```diff
@@ -0,0 +1,33 @@
+
+  List<Widget> toPdf(Map<String, dynamic> data) {
+    return array!
+        .map((child) => child!.buildWidget(data))
```
`array` should be evaluated as an expression against the `data` parameter, which should result in an array (or null). The `map()` function iterates over the resulting array of data. So if you read the part of the task description about the data for the childTemplate, you will need to do something like:

```dart
  final resultArray = // evaluateX(array, dart) <- You need to write this.
  if (resultArray == null) return [];
  return resultArray.mapIndexed((index, childData) {  // mapIndexed() is in the collection package. 
    final dataForChildWidget = {
      ...childData,
      r'$parentData': data,
      r'$index': index,
    };
    return childTemplate!.buildWidget(dataForChildWidget);
  })
  .toList(growable: false);
```

dsyrstad commented on vsc_pdf_template_transformer/lib/src/vsc_pdf_template_transformer_base.dart:
```diff
@@ -27,9 +26,9 @@ class Transformer {
-                jsonDecode(jsonEncode(template['children'][i]['children'][0]))))
-            .toPdf(data));
+        pdf.addPage(
+            TplPage(getWidgetBuilder(template['children'][i]['children'][0]))
```
The ` jsonDecode(jsonEncode(` above should also be removed.

dsyrstad commented on vsc_pdf_template_transformer/lib/models/tpl_page.dart:
```diff
@@ -24,8 +34,16 @@ class TplPage {
+    return pw.Page(
+        pageTheme: pageTheme?.toPdf(data),
+        theme: theme?.toPdf(data),
+        pageFormat: evaluatePageFormat(pageFormat, data),
```
Although there are standard sizes for PdfPageFormat, you can also specify your own page format, most importantly with  margins. I think we need a TplPdfPageFormat class with the PdfPageFormat sizes and margins, plus a `format` attribute which evaluates to a string if it exists. The string would be a3, a4, letter, etc.

dsyrstad commented on vsc_pdf_template_transformer/lib/utils/evaluator.dart:
```diff
@@ -179,3 +179,50 @@ TextOverflow? evaluateTextOverflow(
+  return evaluateEnum(PageOrientation.values, expression, data);
+}
+
+PdfPageFormat? evaluatePageFormat(
```
See comment above about PdfPageFormat.

Commit:
TplPdfPageFormat

dsyrstad commented on vsc_pdf_template_transformer/lib/models/tpl_pdf_page_format.dart:
```diff
@@ -0,0 +1,50 @@
+  Map<String, dynamic> toJson() => _$TplPdfPageFormatToJson(this);
+
+  PdfPageFormat? toPdf(Map<String, dynamic> data) {
+    if (width > 0 && height > 0) {
```
width,  height, and all of the margins need to call `evaluateDouble()` prior to looking at their values.

dsyrstad commented on vsc_pdf_template_transformer/lib/models/tpl_pdf_page_format.dart:
```diff
@@ -0,0 +1,50 @@
+)
+class TplPdfPageFormat {
+  TplPdfPageFormat(
+      {this.width = 0,
```
Don't initialize values in the constructor, and hence don't make the fields `final`. If you initialize them, they will become the defaults if not specified in the JSON. Just have a no-arg constructor: `TplPdfPageFormat()`

dsyrstad replied:
There arguments still needs to be removed.

Commit:
page format and repeater refactor

dsyrstad commented on vsc_pdf_template_transformer/lib/models/tpl_repeater.dart:
```diff
@@ -0,0 +1,43 @@
+    for (int i = 0; i < resultArray.length; i++) {
+      final Map<String, dynamic> dataForChildWidget = {
+        ...resultArray[i],
+        'parentData': data,
```
These should be `r'$parentData'` and `r'$index'` as described on the task.

dsyrstad commented on vsc_pdf_template_transformer/lib/models/tpl_repeater.dart:
```diff
@@ -0,0 +1,43 @@
+
+  Map<String, dynamic> toJson() => _$TplRepeaterToJson(this);
+
+  List<Widget?> toPdf(Map<String, dynamic> data) {
```
You will probably need a common method for other widgets that take a list of children (TplColumn, TplRow, TplMultiPage, etc.) so that they can detect if there is a TplRepeater widget in the list and call this method to add the resulting list of widgets to the array of children. Repeater generates child widgets, so they have to be added somehow. The task has an example of a repeater embedded in a Column. You can use the sample_invoice.json data to test it.

This also should return `List<Widget>` rather than `List<Widget?>`

Commit:
repeater refactor

Commit:
repeater refactor

dsyrstad committed:
Invoice template sample.

dsyrstad commented on vsc_pdf_template_transformer/lib/models/tpl_repeater.dart:
```diff
@@ -0,0 +1,44 @@
+class TplRepeater {
+  String className = 'TplRepeater';
+
+  @WidgetJsonConverter()
```
`array` is data, not a widget.

dsyrstad commented on vsc_pdf_template_transformer/lib/models/tpl_repeater.dart:
```diff
@@ -0,0 +1,44 @@
+  String className = 'TplRepeater';
+
+  @WidgetJsonConverter()
+  List<dynamic>? array;
```
This might be - and probably will be - an expression string until we evaluate it - hence it won't be a list. This should just be `dynamic array`.

dsyrstad commented on vsc_pdf_template_transformer/lib/utils/evaluator.dart:
```diff
@@ -179,3 +185,74 @@ TextOverflow? evaluateTextOverflow(
+      final arr = TplRepeater.fromJson(e).toPdf(data);
+      res.addAll(arr);
+    } else {
+      const widgetClassFromJson = {
+        'TplText': TplText.fromJson,
+        'TplSizedBox': TplSizedBox.fromJson,
+        'TplContainer': TplContainer.fromJson,
+        'TplColumn': TplColumn.fromJson,
+        'TplRow': TplRow.fromJson,
+      };
+
+      final fromJson = widgetClassFromJson[e['className']];
+      final widget = fromJson!(e);
```
`Transformer.getWidgetBuilder()` does this same thing.  We should just call it from here.

dsyrstad commented on vsc_pdf_template_transformer/lib/models/tpl_repeater.dart:
```diff
@@ -0,0 +1,44 @@
+
+  Map<String, dynamic> toJson() => _$TplRepeaterToJson(this);
+
+  List<Widget> toPdf(Map<String, dynamic> data) {
```
I added an invoice template to test this: `vsc_pdf_template_editor/example/assets/test/invoice_tpl_test.json`. Please use it to test the repeater.

Commit:
repeater fix, header widget

dsyrstad committed:
- MultiPage should take any kind of widget for header.
- There was no builder for TplHeader.
- TplPage and TplMultiPage were still not parsing their attirbutes and because of this, TplMultiPage.header just didn't do anything.
- Removed basic_utils dependency.
- Update sample_text.json to be a full template including document and page.
- Added test multi_page_header_tpl_test.json to test TplMultiPage with headers.

dsyrstad commented:
Fixed multiple things:

- MultiPage should take any kind of widget for header. 
- There was no builder for TplHeader. 
- TplPage and TplMultiPage were still not parsing their attirbutes and because of this, TplMultiPage.header just didn't do anything. 
- Removed basic_utils dependency. 
- Update sample_text.json to be a full template including document and page.
- Added test multi_page_header_tpl_test.json to test TplMultiPage with headers.
